v2/service/module/watcher/watcher.go

```diff
@@ -0,0 +1,95 @@
+// Package watcher provides file update notification functionality as a module.
+package watcher
+
+import (
+	"fmt"
+
+	"github.com/fsnotify/fsnotify"
+)
+
+var (
+	ErrMissingWithFilename = fmt.Errorf("watcher.Watcher missing WithFilename option")
+	ErrMissingWithFunc     = fmt.Errorf("watcher.Watcher missing WithFunc option")
+)
+
+type Watcher struct {
+	w        *fsnotify.Watcher
+	stopped  chan struct{}
+	filename *string
+	fn       func() error
+	opts     []Opt
+}
+
+// New creates watcher with given options.
+// WithFilename and WithFunc options are mandatory.
+func New(opts ...Opt) *Watcher {
+	return &Watcher{opts: opts}
+}
+
+func (w *Watcher) Init() error {
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return fmt.Errorf("watcher.Watcher error: %w", err)
+	}
+	w.w = watcher
+	w.stopped = make(chan struct{})
+
+	for _, opt := range w.opts {
+		if err := opt(w); err != nil {
+			return fmt.Errorf("watcher.Watcher Option error: %w", err)
+		}
+	}
+
+	switch {
+	case w.filename == nil:
```

Comment:
With change above this would be just `w.filename == ""` and then we would also catch `WithFilename("")` case.